### PR TITLE
fix: two remotes connect to local

### PR DESCRIPTION
**Pull request by JoanFM**
**Changes introduced**

### CODE REVIEW

File: jina/enums.py [hunk 1]
```diff
@@ -88,15 +88,16 @@ def _from_yaml(cls, constructor, node):
 
 class SchedulerType(BetterEnum):
     LOAD_BALANCE = 0  #: balance the workload between Peas, faster peas get more work
-    ROUND_ROBIN = 1  #: workload are scheduled round-robin manner to the peas, assuming all peas have uniform processing speed.
+    ROUND_ROBIN = 1  # : workload are scheduled round-robin manner to the peas, assuming all peas have uniform
+    # processing speed.
```
Comment (hanxiao) [inline jina/enums.py:92]:
```suggestion
    ROUND_ROBIN = 1  # : workload are scheduled round-robin manner to the peas, assuming all peas have uniform
    # processing speed.
``` 
`#: ` is inline comment

File: jina/parsers/peapods/pea.py [hunk 2]
```diff
@@ -34,7 +34,13 @@ def mixin_pea_parser(parser):
 
     gp.add_argument('--expose-public', action='store_true', default=False,
                     help='If set, expose the public IP address to remote when necessary, by default it exposes'
-                         'private IP address, which only allows accessing under the same network/subnet')
+                         'private IP address, which only allows accessing under the same network/subnet. Important to '
+                         'set this to true when the Pea will receive input connections from remote Peas')
+
+    gp.add_argument('--force-pull-socket-from-remote', action='store_true', default=False,
```
Comment (hanxiao) [inline jina/parsers/peapods/pea.py:40]:
```suggestion
    gp.add_argument('--force-pull-socket-from-remote', action='store_true', default=False,
```

this has to be renamed

Comment (JoanFM) [reply jina/parsers/peapods/pea.py:40]: what name would u suggest?

Comment (hanxiao) [reply jina/parsers/peapods/pea.py:40]: `--force-pull-socket-from-remote` is too casual, we have to be more careful on the naming esp. before 1.0:
1. no need to add `--force` as it is a store_true action
2. `pull-socket` is not accurate, think about it, does it change the push-pull type if you turn on/off this option? no. what is change is the bind/connect: who is bind, who is connect.
3. `remote`? unclear, who is remote? remote to who?

Comment (JoanFM) [reply jina/parsers/peapods/pea.py:40]: I would say force because given the current logic is kind of an exception.

Remote is always a relative concept trying to clarify in the comment.

However I agree is not a very good name but I could not come with a better name

**Review by hanxiao — CHANGES_REQUESTED** [review]
image

too many changes over existing API: the existing remote use case must add this new arg in order to pass the tests?

**Comment by hanxiao** [comment]
If there is a corner case, then add an arg that only for that corner case. In this PR, the newly added arg changes the existing working tests, which introduces extra communication to the community

**Comment by JoanFM** [comment]
> image
> 
> too many changes over existing API: the existing remote use case must add this new arg in order to pass the tests?

Yes and it seems that the tests that were being skipped need that as well.

File: tests/distributed/test_local_flow_local_remote_local/flow.yml [hunk 3]
```diff
@@ -2,6 +2,7 @@
 version: '1'
 with:
   read_only: true
+  public-expose: true
```
Comment (hanxiao) [inline tests/distributed/test_local_flow_local_remote_local/flow.yml:5]: what is `public-expose`? seems irrelevant as the test passing anyway

Comment (JoanFM) [reply tests/distributed/test_local_flow_local_remote_local/flow.yml:5]: yes leftovers from previous hypothesis, my bad

File: jina/enums.py [hunk 4]
```diff
@@ -88,15 +88,15 @@ def _from_yaml(cls, constructor, node):
 
 class SchedulerType(BetterEnum):
     LOAD_BALANCE = 0  #: balance the workload between Peas, faster peas get more work
-    ROUND_ROBIN = 1  #: workload are scheduled round-robin manner to the peas, assuming all peas have uniform processing speed.
+    ROUND_ROBIN = 1  # : workload are scheduled round-robin manner to the peas, assuming all peas have uniform processing speed.
```
Comment (hanxiao) [inline jina/enums.py:91]:
```suggestion
    ROUND_ROBIN = 1  #: workload are scheduled round-robin manner to the peas, assuming all peas have uniform processing speed.
```

File: jina/parsers/peapods/pea.py [hunk 5]
```diff
@@ -34,7 +34,13 @@ def mixin_pea_parser(parser):
...
     gp.add_argument('--expose-public', action='store_true', default=False,
                     help='If set, expose the public IP address to remote when necessary, by default it exposes'
-                         'private IP address, which only allows accessing under the same network/subnet')
+                         'private IP address, which only allows accessing under the same network/subnet. Important to '
+                         'set this to true when the Pea will receive input connections from remote Peas')
+
+    gp.add_argument('--input-socket-pull-connect', action='store_true', default=False,
+                    help='If set, in the case where a Pod is local with respect to the Flow, it forces the incoming '
+                         'socket connection from remote instances to be of type PULL_CONNECT. This is useful when the '
+                         'Pod is not directly reachable from remote and therefore it needs to PULL.')
```
Comment (hanxiao) [inline jina/parsers/peapods/pea.py:43]: remove this arg

Comment (JoanFM) [reply jina/parsers/peapods/pea.py:43]: and how then we make all the tested topologies work?

File: jina/flow/builder.py [hunk 6]
```diff
@@ -92,9 +92,9 @@ def _connect_two_nodes(flow: 'Flow', start_node_name: str, end_node_name: str):
             first_socket_type = SocketType.PUB_BIND
         elif end_node_name == 'gateway':
             first_socket_type = SocketType.PUSH_BIND
-        elif start_node.host != __default_host__ and end_node.host == __default_host__:
+        elif start_node.host != __default_host__ and end_node.host == __default_host__ and end_node.args.input_socket_pull_connect:
```
Comment (hanxiao) [inline jina/flow/builder.py:95]:
```suggestion
        elif start_node.host != __default_host__ and end_node.host == __default_host__ and end_node.args.pod_role !=PodRoleType.JOIN:
```

File: jina/parsers/peapods/pea.py (hunk 5 above)
Comment (hanxiao) [reply jina/parsers/peapods/pea.py:43]: see next comment

Comment (hanxiao) [reply jina/parsers/peapods/pea.py:43]: https://github.com/jina-ai/jina/pull/1809#discussion_r568154650
